Give Tensor a trivial (for now) metaclass _TensorMeta

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56150 Make it illegal to directly construct _TensorBase
* **#56147 Give Tensor a trivial (for now) metaclass _TensorMeta**

This is support of #55686, you can see the broader context of the metaclass in
a more complete PR #56017.  The short story is that in the future I want to
give Tensor a non-trivial metaclass, so to derisk the change first I give it a
trivial metaclass to shake out any bugs that might be caused by it.  The
metaclass shouldn't have any performance impact on Tensor as it only gets
invoked upon subclass creation.

By the way, it was totally not documented how to create metaclasses in the Python
C API, and it took a good bit of trial error to figure it out (and the answer is
now immortalized in https://stackoverflow.com/q/67077317/23845 -- the things
that I got wrong in earlier versions of the PR included setting tp_basicsize
incorrectly, incorrectly setting Py_TPFLAGS_HAVE_GC on the metaclass--you want
to leave it unset so that it inherits, and determining that tp_init is what
actually gets called when you construct a class, not tp_call as another
not-to-be-named StackOverflow question suggests).

Aside: Ordinarily, adding a metaclass to a class is a user visible change, as
it means that it is no longer valid to mixin another class with a different
metaclass. However, because _C._TensorBase is a C extension object, it will
typically conflict with most other metaclasses, so this is not BC breaking.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28028747](https://our.internmc.facebook.com/intern/diff/D28028747)

ezyang committed:
Give Tensor a trivial (for now) metaclass _TensorMeta

This is support of #55686, you can see the broader context of the metaclass in
a more complete PR #56017.  The short story is that in the future I want to
give Tensor a non-trivial metaclass, so to derisk the change first I give it a
trivial metaclass to shake out any bugs that might be caused by it.  The
metaclass shouldn't have any performance impact on Tensor as it only gets
invoked upon subclass creation.

By the way, it was totally not documented how to create metaclasses in the Python
C API, and it took a good bit of trial error to figure it out (and the answer is
now immortalized in https://stackoverflow.com/q/67077317/23845 -- the things
that I got wrong in earlier versions of the PR included setting tp_basicsize
incorrectly, incorrectly setting Py_TPFLAGS_HAVE_GC on the metaclass--you want
to leave it unset so that it inherits, and determining that tp_init is what
actually gets called when you construct a class, not tp_call as another
not-to-be-named StackOverflow question suggests).

Aside: Ordinarily, adding a metaclass to a class is a user visible change, as
it means that it is no longer valid to mixin another class with a different
metaclass. However, because _C._TensorBase is a C extension object, it will
typically conflict with most other metaclasses, so this is not BC breaking.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Give Tensor a trivial (for now) metaclass _TensorMeta"


This is support of #55686, you can see the broader context of the metaclass in
a more complete PR #56017.  The short story is that in the future I want to
give Tensor a non-trivial metaclass, so to derisk the change first I give it a
trivial metaclass to shake out any bugs that might be caused by it.  The
metaclass shouldn't have any performance impact on Tensor as it only gets
invoked upon subclass creation.

By the way, it was totally not documented how to create metaclasses in the Python
C API, and it took a good bit of trial error to figure it out (and the answer is
now immortalized in https://stackoverflow.com/q/67077317/23845 -- the things
that I got wrong in earlier versions of the PR included setting tp_basicsize
incorrectly, incorrectly setting Py_TPFLAGS_HAVE_GC on the metaclass--you want
to leave it unset so that it inherits, and determining that tp_init is what
actually gets called when you construct a class, not tp_call as another
not-to-be-named StackOverflow question suggests).

Aside: Ordinarily, adding a metaclass to a class is a user visible change, as
it means that it is no longer valid to mixin another class with a different
metaclass. However, because _C._TensorBase is a C extension object, it will
typically conflict with most other metaclasses, so this is not BC breaking.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

albanD commented on test/test_jit.py:
```diff
@@ -7567,7 +7567,7 @@ def test_type_call_in_script(self):
             return type(x)
 
-        with self.assertRaisesRegex(RuntimeError, "value of type type"):
+        with self.assertRaisesRegex(RuntimeError, "value of type _TensorMeta"):
```
I'm not sure to understand what this means? Is it that `type(type(my_tensor))` is not `_TensorMeta` instead of `type`?

ezyang replied:
The error here is raised by TorchScript where it complains that it doesn't know how to deal with values like "torch.Tensor" (which, prior to this PR, was a "type"). After this, it's a "_TensorMeta", so the error message wobbles (but not in a substantive way)

albanD replied:
Ok, should this be marked as BC-breaking then?

ezyang replied:
We can conservatively mark it as BC breaking but I currently don't think anyone will actually be broken by it in practice.

albanD replied:
It mostly depends if we want this to appear in the release notes and if so how.
I feel like it is nice to mention it at least no?

ezyang replied:
Yeah, talking about it in internals sgtm

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -809,8 +809,70 @@ static PyMethodDef extra_methods[] = {
+   of Python objects defined in other libraries to be used in static
+   initializers here.  The DEFERRED_ADDRESS macro is used to tag the slots where
+   such addresses appear; the module init function must fill in the tagged slots
+   at runtime.  The argument is for documentation -- the macro ignores it.
```
super nice comment! Thanks!

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -809,8 +809,70 @@ static PyMethodDef extra_methods[] = {
+}
+
+PyTypeObject THPVariableMetaType = {
+  PyVarObject_HEAD_INIT(DEFERRED_ADDRESS(&PyType_Type), 0)
```
Is this `PyVarObject_HEAD_INIT` doing anything non-trivial? Because it would do it before the `PyType_Type` is actually populated during the init here no?

Actually that field is not updated in the init. So it shouldn't be a DEFERRED_ADDRESS ?

ezyang replied:
It gets automatically initialized by `tp_base`. https://github.com/python/cpython/blob/v3.7.0/Modules/xxsubtype.c#L263-L266

albanD reviewed (commented):
Looks good, just minor questions

albanD reviewed (approved):
LGTM
Maybe BC-breaking, for documentation purposes, but that's it.

ezyang committed:
Update on "Give Tensor a trivial (for now) metaclass _TensorMeta"


This is support of #55686, you can see the broader context of the metaclass in
a more complete PR #56017.  The short story is that in the future I want to
give Tensor a non-trivial metaclass, so to derisk the change first I give it a
trivial metaclass to shake out any bugs that might be caused by it.  The
metaclass shouldn't have any performance impact on Tensor as it only gets
invoked upon subclass creation.

By the way, it was totally not documented how to create metaclasses in the Python
C API, and it took a good bit of trial error to figure it out (and the answer is
now immortalized in https://stackoverflow.com/q/67077317/23845 -- the things
that I got wrong in earlier versions of the PR included setting tp_basicsize
incorrectly, incorrectly setting Py_TPFLAGS_HAVE_GC on the metaclass--you want
to leave it unset so that it inherits, and determining that tp_init is what
actually gets called when you construct a class, not tp_call as another
not-to-be-named StackOverflow question suggests).

Aside: Ordinarily, adding a metaclass to a class is a user visible change, as
it means that it is no longer valid to mixin another class with a different
metaclass. However, because _C._TensorBase is a C extension object, it will
typically conflict with most other metaclasses, so this is not BC breaking.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28028747](https://our.internmc.facebook.com/intern/diff/D28028747)

[ghstack-poisoned]
